Design Document for SRE Challenge

dereknorrbom opened the pull request:
Design Document for SRE Challenge

dereknorrbom committed:
add rfd 0

add rfd 0

evanfreed commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+
+### Security
+
+The chosen cipher suites for mTLS will be:
+- TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384
+- TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384
+- TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256
+- TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
```
Any reason against using TLS 1.3?

dereknorrbom replied:
I've been fortunate enough to work in environments where I only needed to set the TLS version without memorizing all the ciphers. I initially went with TLS 1.2 based on common recommendations. However, TLS 1.3 is the better choice for stronger security and faster performance. I'll update the design document to specify TLS 1.3.

evanfreed commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running.
+- **GET /replica-count**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count**: Sets the replica count of the specified deployment.
+- **GET /deployments**: Lists all deployments in the specified namespace.
```
How will the API support specifying the namespace?

dereknorrbom replied:
I planned to hardcode the namespace value based on the challenge documentation's suggestion to "_Use hardcoded values as much as possible and simply add TODO items showing your thinking._"  To support specifying the namespace, I can modify the API design to accept it as a query parameter.

evanfreed commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+
+### Extending to Multiple Target Kubernetes Clusters
+
+While the current implementation targets a single Kubernetes cluster, extending it to support multiple clusters involves several key considerations:
```
Just clarifying since you asked about this in the Slack channel.  Is this something you plan on implementing? If not, maybe include a comment that this is how you would go about this.

dereknorrbom replied:
I do not plan on supporting multiple clusters within the scope of this project. I will update the document to make it clear that the section on multiple clusters is just an explanation of how I could support it if it were part of the project scope.

taraspos commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running.
+- **GET /replica-count**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count**: Sets the replica count of the specified deployment.
```
Similar to the question below:
- How are we going to provide provide deployment name in the request?
- What kind of output should we expect?

dereknorrbom replied:
I planned to hardcode the deployment name based on the challenge documentation's suggestion to "_Use hardcoded values as much as possible and simply add TODO items showing your thinking._" To support specifying the deployment name, I will modify the API design to accept it as a query parameter.  

The API will return JSON responses for all endpoints, I will update the design document to reflect this.

taraspos commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+clean:
+	@echo "Cleaning up..."
+	rm -rf ./certs
+	kind delete cluster --name k8s-deployment-scaler
```
`clean` is fairy common command to run and not sure if I would want it to destroy the cluster every time.
What do you think about separate command something like `teardown`?

dereknorrbom replied:
I agree that having the clean command destroy the cluster may not be ideal. I'll separate the cluster and Docker teardown into a different command.

taraspos commented on rfd/0000-sre-challenge.md:
````diff
@@ -0,0 +1,213 @@
+   ```
+
+2. **Run the Setup Script**:
+   A `setup.sh` script will be provided to automate the detection and installation of the required dependencies (Docker, KIND, Helm, kubectl (Kubernetes command-line tool)).
````
What kind of OS will your script target?

Installation for those tools might be slightly different for MacOS/Linux/Windows

dereknorrbom replied:
The setup.sh script will target x86 64-bit Linux and MacOS environments, as specified in the challenge documentation.  Windows is out of scope for this project, so the script will not include Windows-specific installation instructions.

taraspos commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,213 @@
+
+### Delivery
+
+The API will be delivered as a Docker image, which can be deployed to any Kubernetes cluster using Helm charts. This ensures that the API can be easily deployed and managed in different environments.
```
What kind of kubernetes resources are you planning to deploy with the helm chart?

dereknorrbom replied:
The Helm chart will deploy the following Kubernetes resources:

1. Deployment: This will manage the deployment of the API Docker container, ensuring the specified number of replicas are running.
2. Service: This will expose the API deployment within the Kubernetes cluster, allowing it to be accessed by other services or through port forwarding.
3. ClusterRole and ClusterRoleBinding: These will grant cluster-wide permissions necessary for the API to interact with Kubernetes resources.
4. Role and RoleBinding: These will grant namespace-specific permissions for the API to interact with Kubernetes resources.

dereknorrbom committed:
Update Design Document

Update the design document to reflect feedback from the initial pull request comments.

dereknorrbom commented:
The design document has been updated to reflect the comment questions/feedback. @evanfreed @taraspos

jimbishopp commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,264 @@
+The API will provide the following endpoints:
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running and Kubernetes connectivity is verified.
+- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
+- **POST /set-replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.
```
These two endpoints are very similar. They both accept the same parameters and operate on replica count. What is the reason for defining two endpoints rather than one endpoint with different HTTP verbs?

dereknorrbom replied:
In pedantic REST, the endpoints would use the same name with different verbs. I went with descriptive names for clarity.

Proposed Update:
- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
- **POST /replica-count?namespace={namespace}&deployment={deployment}**: Sets the replica count of the specified deployment.

jimbishopp commented on rfd/0000-sre-challenge.md:
````diff
@@ -0,0 +1,264 @@
+   }
+   ```
+
+- **POST /set-replica-count?namespace=default&deployment=my-deployment**
+   ##### Request Body
+   ```json
+   {
+      "replicas": 3
+   }
+   ```
+   ##### Response
+   ```json
+   {
+      "message": "Replica count updated to: 3"
+   }
+   ```
````
The response message reads a bit like a log message. What do you think about returning the same message as returned from the get replica count endpoint?

dereknorrbom replied:
Returning the same message as the GET replica count endpoint makes sense for consistency. I'll update the response to match the GET endpoint.

jimbishopp commented on rfd/0000-sre-challenge.md:
````diff
@@ -0,0 +1,264 @@
+   }
+   ```
+
+- **GET /deployments?namespace=default**
+   ##### Response
+   ```json
+   {
+      "deployments": ["my-deployment"]
+   }
+   ```
````
Will the deployment strings in the array include namespace?

dereknorrbom replied:
Currently, the design does not include the namespace in the deployment strings. Including the namespace can provide more clarity, especially when working with multiple namespaces.

Proposed update:

```json
{
  "deployments": ["default/my-deployment"]
}
```

dereknorrbom committed:
Update Design Document

Update document based on pull request comment feedback.
- Use the same endpoint for replica-count GET/POST
- replica-count POST response to match GET response
- Include the namespace in the deployment response

dereknorrbom commented:
The design document has been updated to reflect the comment questions/feedback. @jimbishopp @evanfreed @taraspos

logand22 commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,264 @@
+The API will provide the following endpoints:
+
+- **GET /healthz**: Health check endpoint that returns "OK" if the server is running and Kubernetes connectivity is verified.
+- **GET /replica-count?namespace={namespace}&deployment={deployment}**: Returns the current replica count of the specified deployment.
```
Are these query parameters required?

dereknorrbom replied:
They were hardcoded originally per "_Use hardcoded values as much as possible and simply add TODO items showing your thinking_", and updated to utilize query parameters based on this comment https://github.com/dereknorrbom/k8s-deployment-scaler/pull/1#discussion_r1644183083.  I plan on using the hardcoded values as fallbacks when a deployment/namespace is not provided in the request.

logand22 replied:
Sorry, I meant can you leave the namespace and deployment parameters empty and receive all replica counts? Or are you saying your application will only support monitoring a single deployment?

logand22 commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,264 @@
+
+Each endpoint will be secured using mutual TLS (mTLS) to ensure secure communication between the client and the server.
+
+#### Example JSON Responses
```
Can you provide some examples of error responses.

dereknorrbom replied:
Yes, I will update the design document to include example error responses.

logand22 commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,264 @@
+	kind delete cluster --name k8s-deployment-scaler
+	docker rmi k8s-deployment-scaler:latest
+
+# Generate certificates by copying them from the Docker image to the local certs directory
```
But where do these certs originate from? Are they built into the container? If so can you provide more information on this.

dereknorrbom replied:
For simplicity, the certs are self-signed and generated within the container.

logand22 commented on rfd/0000-sre-challenge.md:
```diff
@@ -0,0 +1,264 @@
+##### Implementation Details:
+1. **In-Memory Cache**: Use an in-memory map to store the replica counts.
+2. **Synchronization**: Use a `sync.RWMutex` to manage concurrent access to the cache.
+3. **Kubernetes Informers**: Use Kubernetes informers to watch for changes to deployments and update the cache accordingly.
```
Just to clarify, you'll create an in-memory cache and then use `informers` to populate that cache? If so, is there a reason you don't use the cache that's built into the informer?

dereknorrbom replied:
It was an oversight on my part. I'll update the design to use the informer's built-in cache.

evanfreed reviewed (approved):
@dereknorrbom since you have the approvals necessary feel free to merge this when ready and move onto the implementation part of the challenge. 

As a reminder please focus on the `Server` components for the first PR (if there's some simple build targets to support it that's fine).
